### PR TITLE
[MERGE] stock: picking: button_validate multi

### PR DESCRIPTION
**Pull request by sle-odoo**
preliminary work for
- task 1938108 (which is adding a pre action done wizard)
- task 1938656 (which is adding another pre action done wizard)
- task 2069646 (which will rework all pre action done wizard to select to pickings to apply)

### CODE REVIEW

File: addons/stock_picking_batch/models/stock_picking_batch.py [hunk 1]
```diff
@@ -67,43 +67,8 @@ def done(self):
...
-                'views': [(view.id, 'form')],
-                'view_id': view.id,
-                'target': 'new',
-                'res_id': wiz.id,
-                'context': self.env.context,
-            }
-        if picking_to_backorder:
-            return picking_to_backorder.action_generate_backorder_wizard()
-        return True
+        return self.picking_ids.button_validate()
```
Comment (sle-odoo) [inline addons/stock_picking_batch/models/stock_picking_batch.py]: if the user cancel the backorer the batch picking is still marked as done.. fuck

File: addons/stock/models/stock_picking.py [hunk 2]
```diff
@@ -820,28 +819,38 @@ def button_validate(self):
...
-                'type': 'ir.actions.act_window',
-                'view_mode': 'form',
-                'res_model': 'stock.immediate.transfer',
-                'views': [(view.id, 'form')],
-                'view_id': view.id,
-                'target': 'new',
-                'res_id': wiz.id,
-                'context': self.env.context,
-            }
+        # Run the pre-validation wizards. Processing a pre-validation wizard should work on the
```
Comment (amoyaux) [inline addons/stock/models/stock_picking.py]: care with the sanity check with multiple picking that have different picking type (server action for e.g.)

File: addons/stock/wizard/stock_backorder_confirmation.py [hunk 3]
```diff
@@ -22,10 +22,18 @@ def _process(self, cancel_backorder=False):
                                          precision_rounding=move.product_uom.rounding) > 0:
                             moves_to_log[move] = (move.quantity_done, move.product_uom_qty)
                     pick_id._log_less_quantities_than_expected(moves_to_log)
-            confirmation.pick_ids.with_context(cancel_backorder=cancel_backorder)._action_done()
+
+        pickings_to_validate = self.env.context.get('button_validate_picking_ids')
+        if pickings_to_validate:
+            pickings_to_validate = self.env['stock.picking'].browse(pickings_to_validate).with_context(skip_backorder=True)
+            if cancel_backorder:
+                pickings_to_validate = pickings_to_validate.with_context(picking_ids_not_to_backorder=self.pick_ids.ids)
```
Comment (amoyaux) [inline addons/stock/wizard/stock_backorder_confirmation.py]: could use the same context key for picking_ids_not_to_backorder and skip_backorder

File: addons/stock/wizard/stock_immediate_transfer.py [hunk 4]
```diff
@@ -25,13 +23,9 @@ def process(self):
...
-        if pick_to_do:
-            pick_to_do._action_done()
-        if pick_to_backorder:
-            return pick_to_backorder.action_generate_backorder_wizard()
-        return False
+
+        pickings_to_validate = self.env.context.get('button_validate_picking_ids')
+        if pickings_to_validate:
+            return self.env['stock.picking'].browse(pickings_to_validate).with_context(skip_immediate=True).button_validate()
+        return True
```
Comment (sle-odoo) [inline addons/stock/wizard/stock_immediate_transfer.py]: if the user hits cancel, nothing is processed

Comment (sle-odoo) [reply addons/stock/wizard/stock_immediate_transfer.py]: > added a skip button

File: addons/stock_picking_batch/models/stock_picking_batch.py (hunk 1 above)
Comment (sle-odoo) [reply addons/stock_picking_batch/models/stock_picking_batch.py]: don't care atm

File: addons/stock/models/stock_picking.py (hunk 2 above)
Comment (sle-odoo) [reply addons/stock/models/stock_picking.py]: fixed

File: addons/stock/wizard/stock_backorder_confirmation.py (hunk 3 above)
Comment (sle-odoo) [reply addons/stock/wizard/stock_backorder_confirmation.py]: don't think so

**Pull request closed without merging**